zarr/storage.py

```diff
@@ -2224,27 +2224,25 @@ def __iter__(self):
             list_blobs_prefix = self.prefix + '/'
         else:
             list_blobs_prefix = None
-        for blob in self.client.list_blobs(self.container, list_blobs_prefix):
+        for blob in self.client.list_blobs(list_blobs_prefix):
             yield self._strip_prefix_from_path(blob.name, self.prefix)
 
     def __len__(self):
         return len(self.keys())
 
     def __contains__(self, key):
         blob_name = self._append_path_to_prefix(key)
-        if self.client.exists(self.container, blob_name):
-            return True
-        else:
-            return False
+        return self.client.get_blob_client(blob_name).exists()
 
     def listdir(self, path=None):
-        from azure.storage.blob import Blob
+        # from azure.storage.blob import Blob
         dir_path = normalize_storage_path(self._append_path_to_prefix(path))
         if dir_path:
             dir_path += '/'
         items = list()
-        for blob in self.client.list_blobs(self.container, prefix=dir_path, delimiter='/'):
-            if type(blob) == Blob:
+        for blob in self.client.list_blobs(name_starts_with=dir_path):
+            # items.append(self._strip_prefix_from_path(blob.name, dir_path))
+            if '/' not in blob.name:  # what is this doing?
                 items.append(self._strip_prefix_from_path(blob.name, dir_path))
             else:
                 items.append(self._strip_prefix_from_path(
```

Comment:
@tjcrone - I'm pretty confused here. Any points on the `listdir` method would be much appreciated.

Reply:
https://github.com/zarr-developers/zarr-python/blob/e3cdd1a50e1047304be2c91a017da6362f8df533/zarr/storage.py#L8

Reply:
The `listdir` method which was implemented using the old version of the azure storage library, used the `list_blobs` method in that old library. That method in the old library had a [delimiter](https://docs.microsoft.com/en-us/rest/api/storageservices/list-blobs#uri-parameters) parameter, which helps in listing directories much easier. For example if we have two directories in the container, say `foo/` and `bar/` each of which could have hundreds of thousands of blobs, we would not want to list all of them in the `list_blobs` operations. Instead if you specify `delimiter='/'`, you would only be returned `['foo/', 'bar/']` in the `list_blobs` operation. The new version of the azure storage library seems not to have the `delimiter` option in the `list_blobs` function. Instead there is a function [walk_blobs](https://azuresdkdocs.blob.core.windows.net/$web/python/azure-storage-blob/12.5.0/azure.storage.blob.html#azure.storage.blob.ContainerClient.walk_blobs) which has this parameter. I think that function would be more appropriate here instead of `list_blobs`. I think just replacing `self.client.list_blobs(name_starts_with=dir_path)` with `self.client.walk_blobs(name_starts_with=dir_path, delimiter='/')`, should work.